src/ot_propagation.erl

```diff
@@ -20,25 +20,46 @@
 -export([http_inject/1,
          http_extract/1]).
 
--type data() :: term().
+-type extractor(T) :: {fun((T, ot_ctx:context_manager(), ot_ctx:namespace(), {fun(), term()}) -> ok), term()} |
+                      {fun((T, ot_ctx:context_manager(), ot_ctx:namespace(), ot_ctx:key(), {fun(), term()}) -> ok), term()}.
+-type injector(T) :: {fun((T, ot_ctx:context_manager(), ot_ctx:namespace(), {fun(), term()}) -> T), term()} |
+                     {fun((T, ot_ctx:context_manager(), ot_ctx:namespace(), ot_ctx:key(), {fun(), term()}) -> T), term()}.
 
--type extractor() :: fun((data()) -> ok).
--type injector() :: fun((data()) -> data()).
 -type http_headers() :: [{iodata(), iodata()}].
 
--type http_injector() :: fun((http_headers()) -> http_headers()).
--type http_extractor() :: fun((http_headers()) -> ok).
+-type http_injector() :: injector(http_headers()).
+-type http_extractor() :: extractor(http_headers()).
 
--export_type([extractor/0,
-              injector/0,
+-export_type([extractor/1,
+              injector/1,
               http_injector/0,
               http_extractor/0,
               http_headers/0]).
 
 http_inject(Headers) ->
-    Fun = opentelemetry:get_http_injector(),
-    Fun(Headers).
+    Injectors = opentelemetry:get_http_injector(),
+    run_injectors(Headers, Injectors).
 
 http_extract(Headers) ->
-    Fun = opentelemetry:get_http_extractor(),
-    Fun(Headers).
+    Extractors = opentelemetry:get_http_extractor(),
+    run_extractors(Headers, Extractors).
+
+run_extractors(Headers, Extractors) ->
```

Comment:
just to be sure, we'll always just be running the locally-defined callbacks with the exact tuple values (3-tuples and 4-tuples) there for the arguments? Would it have made sense to instead specify them as accepting a term for opts in a second argument or something?

Reply:
Yes. It takes the context manager, where to insert in the context (namespace with or without an additional key) and how to decode/encode.